65 reweighting Scottish EPC records

crispy-wonton opened the pull request:
Fixes #65, #88 and #89 
---

# Description

Add reweighting of Scottish EPC records into pipeline. Summary of changes:
- `config/base.yaml` - add new Scottish census datasets for reweighting 
- `get_target.py` - add new getters to get Scottish target data for reweighting. (for property type and tenure but not build year)
- `prepare_target.py` - use new getter to prepare target dataset for reweighting
- `prepare_sample.py` - changes to shorten category names in property type feature
- `run_compute_epc_weights.py` - update run script to reweight Scotland with 2 features and England and Wales with 3
- `reweight_epc.py` - add TODO note

NB: information about data sources will be added to the README in a separate PR.

# Instructions for Reviewer

To run code:
`python -i asf_heat_pump_suitability/pipeline/run_scripts/run_compute_epc_weights.py --epc s3://asf-daps/lakehouse/processed/epc/old/deduplicated/processed_dedupl-0.parquet -y 2023 -q 4`

NB: The whole pipeline will take a few hours to run so please could you just test to make sure it works for Scotland and then you can interrupt the run.

Here are the latest outputs of the pipeline which you can load and check instead of running it all through:
Weights: s3://asf-heat-pump-suitability/outputs/2023Q4/20241129_2023_Q4_EPC_weights.parquet
Stats: s3://asf-heat-pump-suitability/outputs/2023Q4/20241129_2023_Q4_EPC_weights_stats.parquet

Please pay special attention to ...
- Reweighting has been applied correctly to each nation in the code. I.e. Scotland is reweighted appropriately on property_type and tenure features only, whereas England and Wales should also be reweighted on build_year
- Joins to make sure they are correct and we are not accidentally losing data we should keep. My intention is to retain all the rows that get assigned a weight, but not to keep the unweighted ones. The idea is to join this dataset back onto the full EPC dataset in a separate processing script.
- The preprocessing steps in `run_compute_epc_weights.py` before reweighting occurs to ensure they are applied in the appropriate order before and after filtering for country
- Feel free to suggest places where we can add test/assertions to ensure the pipeline is executing the expected behaviours. I'd like to look at this properly myself later and add some in a separate PR.

# Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [ ] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [ ] Appropriate information has been added to `README`s
- [x] I have explained this PR above
- [x] I have requested a code review

crispy-wonton committed:
add getters to load scottish census tenure and property type datasets for reweighting
remove unused functions from get_target.py

crispy-wonton committed:
add new file paths to config for tenure and property type from scottish census

crispy-wonton committed:
update property type category names in prepare_sample.py

crispy-wonton committed:
update functions in prepare_target.py to use EW and Scottish data for reweighting for tenure and property type

crispy-wonton committed:
update run_compute_epc_weights to allow reweighting of Scotland with 2 features and England and Wales with 3

crispy-wonton committed:
Merge branch 'dev' into 65_scottish_reweighting

crispy-wonton committed:
update tenure categories in target data to align with EPC in get_target.py

crispy-wonton committed:
fix bug in get_target.py
4 Scottish DataZones have 0 values for all property types and all tenure types. These data cannot be used for reweighting

crispy-wonton committed:
add TODO note to reweight_epc.py to prevent erroring out of pipeline if sample reduced to len 0

crispy-wonton committed:
Merge branch 'dev' into 65_scottish_reweighting

crispy-wonton committed:
remove epc use_cols from base config

crispy-wonton committed:
remove unused property_type raw getter from get_datasets.py and base config

crispy-wonton committed:
add LSOA to weights dict in run_compute_epc_weights.py to retain LSOA information for dummy rows

helloaidank commented on asf_heat_pump_suitability/config/base.yaml:
```diff
@@ -2,8 +2,8 @@ data_source:
   gb_ons_postcode_dir_url: "https://www.arcgis.com/sharing/rest/content/items/487a5ba62c8b4da08f01eb3c08e304f6/data" # Aug 2023 data
   gb_ons_postcode_dir_file_path: "Data/ONSPD_AUG_2023_UK.csv" # Aug 2023 data
   UK_ons_postcode_dir: "s3://asf-heat-pump-suitability/source_data/ONSPD_AUG_2023_UK.csv"
-  EW_census_housing_characteristics: "s3://asf-heat-pump-suitability/source_data/2021census_Mar2023update_housing_characteristics_E_W.xlsx" # 2021 census, Mar 2023 update
```
Just for my understanding, why did we get rid of this in case I've misunderstood? It's no longer relevant?

crispy-wonton replied:
Yes, I searched in the repo for references to it and it isn't used anywhere so I deleted it.

helloaidank commented on asf_heat_pump_suitability/config/base.yaml:
```diff
@@ -48,25 +49,6 @@ data_source:
   EW_inspire_url: "https://use-land-property-data.service.gov.uk/datasets/inspire/download"
   S_scottish_gov_DZ2011_boundaries: "s3://asf-heat-pump-suitability/source_data/2014_Scottish_Government_DataZoneBoundaries_2011_S/SG_DataZone_Bdry_2011.shp"
   S_NRScotland_dwellings: "s3://asf-heat-pump-suitability/source_data/June2024_NRScotland_households_and_dwellings_S.xlsx"
-usecols:
```
Where has this gone? or we just don't use it anymore?

crispy-wonton replied:
This was originally used in `run_compute_epc_weights.py` to load the EPC data with these columns. I thought it was easier and more readable to change the columns in the script itself, so it's moved there now. Let me know if you think it should be returned.

helloaidank replied:
No, no, was just genuinely curious!

helloaidank commented on asf_heat_pump_suitability/config/base.yaml:
```diff
@@ -2,8 +2,8 @@ data_source:
   gb_ons_postcode_dir_url: "https://www.arcgis.com/sharing/rest/content/items/487a5ba62c8b4da08f01eb3c08e304f6/data" # Aug 2023 data
```
Potentially we can clean up all these links up with something like
```
data_source:
  # ONS Postcode Directory
  gb_ons_postcode_dir:
    url: "https://www.arcgis.com/sharing/rest/content/items/487a5ba62c8b4da08f01eb3c08e304f6/data" # Aug 2023 data
    file_path: "Data/ONSPD_AUG_2023_UK.csv" # Aug 2023 data
    s3_path: "s3://asf-heat-pump-suitability/source_data/ONSPD_AUG_2023_UK.csv"

  # Census Data
  census:
    EW_tenure: "s3://asf-heat-pump-suitability/source_data/2021census_2023Mar_tenure_E_W.csv"
    S_tenure: "s3://asf-heat-pump-suitability/source_data/2022_Scotlands_census_tenure_S.csv"
    EW_number_of_rooms: "s3://asf-heat-pump-suitability/source_data/2021census_Mar2023_number_of_rooms_E_W.csv"
    EW_number_of_households: "s3://asf-heat-pump-suitability/source_data/2021_vMar2023_census_numberofhouseholds_EW.csv"
    EW_land_area: "s3://asf-heat-pump-suitability/source_data/2021_vMar2021_census_landareaKM_EW.csv"
    EW_accommodation_type: "s3://asf-heat-pump-suitability/source_data/2021census_Mar2023_accommodation_type_E_W.csv"
    S_accommodation_type: "s3://asf-heat-pump-suitability/source_data/2022_Scotlands_census_accommodation_type_S.csv"

  # Dwelling Age Data
  dwelling_age:
    EW_cdrc: "s3://asf-heat-pump-suitability/source_data/2015cdrc_dwelling_ages_E_W.csv"
    EW_cdrc_la: "s3://asf-heat-pump-suitability/source_data_minor_edits/2015cdrc_dwelling_ages_E_W_per_la_02.parquet"

  # Garden Space Access
  garden_space_access:
    GB_ons: "s3://asf-heat-pump-suitability/source_data/ONS_Apr2020_access_to_garden_space.xlsx"

  # UPRN Data
  uprn:
    GB_osopen: "s3://asf-heat-pump-suitability/source_data/osopenuprn_202405_csv.zip"

  # Local Authority Districts Boundaries
  lad_bounds:
    UK_ons: "s3://asf-heat-pump-suitability/source_data/Local_Authority_Districts_December_2023_Boundaries_UK_BFE_-2600600853110041429/LAD_DEC_2023_UK_BFE.shp"

  # Inspire Land Extent
  inspire_land_extent:
    EW: "s3://asf-heat-pump-suitability/source_data/inspire_ew/"
    S: "s3://asf-heat-pump-suitability/source_data/inspire_scotland/"

  # Microsoft Building Footprint Links
  microsoft_building_footprint:
    global: "https://minedbuildings.z5.web.core.windows.net/global-buildings/dataset-links.csv"

  # Conservation Areas
  conservation_areas:
    E_historic_england: "s3://asf-heat-pump-suitability/source_data/Aug2024_historic_england_conservation_areas_E.geojson"
    W_welsh_gov: "s3://asf-heat-pump-suitability/source_data/2022_welsh_gov_building_conservation_areas_W.gpkg"

  # Off Gas Grid
  off_gas_grid:
    UK_spa: "s3://asf-heat-pump-suitability/source_data/2024_vMar2024_SPA_offgaspostcode_UK.xlsx"

  # Listed Buildings
  listed_buildings:
    E_historic_england: "s3://asf-heat-pump-suitability/source_data/Jun2024_vJul2024_HistoricEngland_listedbuilding_E.gpkg"
    W_cadw: "s3://asf-heat-pump-suitability/source_data/May2024_vMay2024_Cadw_listedbuilding_W.gpkg"
    S_scottish_gov: "s3://asf-heat-pump-suitability/source_data/lb_scotland/Listed_Buildings.shp"

  # LSOA to LAD Lookup
  lsoa_lad_lookup:
    EW_ons: "s3://asf-heat-pump-suitability/source_data/2021_vApr2023_ons_lsoa_to_lad_lookup_EW.csv"

  # LSOA Boundaries
  lsoa_bounds:
    EW: "s3://asf-heat-pump-suitability/source_data/Lower_layer_Super_Output_Areas_2021_EW_BFE_V9_-9107090204806789093/LSOA_2021_EW_BFE_V9.shp"

  # Points of Interest Locations
  poi_locations:
    UK: "s3://asf-heat-pump-suitability/source_data/poi_uk.gpkg"

  # Grid Operators Data
  grid_operators:
    E_ENW:
      dfes_primaries: "s3://asf-heat-pump-suitability/source_data/grid_operators/ENW/dfes-2023-primary-data0.parquet"
      ndp_headroom: "s3://asf-heat-pump-suitability/source_data/grid_operators/ENW/ndp-pry-bsp-headroom.parquet"
      ndp_voronoi: "s3://asf-heat-pump-suitability/source_data/grid_operators/ENW/ndp-pry-voronoi.parquet"
    E_NPg:
      heatmap: "s3://asf-heat-pump-suitability/source_data/grid_operators/NPg/heatmapdemanddata.parquet"
      ndp_demand: "s3://asf-heat-pump-suitability/source_data/grid_operators/NPg/npg_ndp_demand_headroom.parquet"
    S_SPEN:
      spd_substations: "s3://asf-heat-pump-suitability/source_data/grid_operators/SPEN/distributed-generation-sp-distribution-heat-maps-spd-primary-substations.parquet"
      spd_polygons: "s3://asf-heat-pump-suitability/source_data/grid_operators/SPEN/ndp-spd-primary-substation-polygons.parquet"
    W_SPEN:
      spm_substations: "s3://asf-heat-pump-suitability/source_data/grid_operators/SPEN/distributed-generation-sp-manweb-heat-maps-spm-primary-substations.parquet"
      spm_polygons: "s3://asf-heat-pump-suitability/source_data/grid_operators/SPEN/ndp-spm-primary-group-polygons.parquet"
    ES_SSEN:
      demand: "s3://asf-heat-pump-suitability/source_data/grid_operators/SSEN/demand-heat-map-update-feb-24.xlsx - PS.csv"
    S_SSEN:
      sepd_bounds: "s3://asf-heat-pump-suitability/source_data/grid_operators/SSEN/SEPD Primary Sub Boundaries"
    E_SSEN:
      shepd_bounds: "s3://asf-heat-pump-suitability/source_data/grid_operators/SSEN/SHEPD Primary Sub Boundaries"
    E_UKPN:
      primaries: "s3://asf-heat-pump-suitability/source_data/grid_operators/UKPN/ukpn_primary_postcode_area.parquet"
    EW_WPD:
      capacity: "s3://asf-heat-pump-suitability/source_data/grid_operators/WPD/wpd-network-capacity-map.csv"
    E_WPD:
      east_midlands_bounds: "s3://asf-heat-pump-suitability/source_data/grid_operators/WPD/east-midlands-primary.gpkg"
      south_west_bounds: "s3://asf-heat-pump-suitability/source_data/grid_operators/WPD/south-west-primary.gpkg"
      west_midlands_bounds: "s3://asf-heat-pump-suitability/source_data/grid_operators/WPD/west-midlands-primary.gpkg"
    W_WPD:
      south_wales_bounds: "s3://asf-heat-pump-suitability/source_data/grid_operators/WPD/south-wales-primary.gpkg"

  # World Heritage Sites
  world_heritage_sites:
    S_historic_environment_scotland: "s3://asf-heat-pump-suitability/source_data/WHS/World_Heritage_Sites.shp"

  # Inspire URLs
  inspire_urls:
    S_ros: "https://ros.locationcentre.co.uk/inspire/"
    EW: "https://use-land-property-data.service.gov.uk/datasets/inspire/download"

  # Scottish Government Data Zone Boundaries
  scottish_gov_dz_boundaries:
    S: "s3://asf-heat-pump-suitability/source_data/2014_Scottish_Government_DataZoneBoundaries_2011_S/SG_DataZone_Bdry_2011.shp"

  # NRS Scotland Dwellings
  nrs_scotland_dwellings:
    S: "s3://asf-heat-pump-suitability/source_data/June2024_NRScotland_households_and_dwellings_S.xlsx"
```

crispy-wonton replied:
This is a great idea. Thank you for taking the time to do this! We should definitely clean the data sources in here as it is getting messy and confusing to find things. I especially like that you came up with the ["EW"] and ["S"] keys!

I will suggest we create a separate issue and PR for this as this would require a lot of refactoring. Although it should be easy to do, we have to do it carefully to ensure that all the places where the `config` is called is updated with the appropriate keys.

We also will have to update the `config/README` `config key` column.

crispy-wonton replied:
I have opened a new issue #92 for this :) thank you!

helloaidank commented on asf_heat_pump_suitability/getters/get_target.py:
```diff
@@ -1,8 +1,7 @@
-from asf_heat_pump_suitability.getters import base_getters
 
 
 def get_df_target_nrooms() -> pl.DataFrame:
```
Is there a reason why we don't include Scotland in the `get_df_target_nrooms()` function?

crispy-wonton replied:
Good spot thank you! At the moment we don't use the `get_df_target_nrooms()` function at all but I left it in the code from when I originally wrote it in case we wanted to implement it at a later point. I will add a #  TODO note to flag that Scotland would need to be added if we do use it.

helloaidank replied:
Makes sense

helloaidank commented on asf_heat_pump_suitability/getters/get_target.py:
```diff
@@ -78,50 +90,128 @@ def get_df_target_property_type_uncensored() -> pl.DataFrame:
         )
-        .drop(["Area Name"])
-        .rename({"Area Code": "lsoa"})
+        .rename({"Type of accomodation": "lsoa"})
```
I might be misunderstanding this, but why are we renaming this from "Type of accommodation" to "lsoa"?

crispy-wonton replied:
Thanks for flagging this because it is weird! I double checked it, and it is correct that we rename this to LSOA.

Basically, the original csv file is poorly formatted with a lot of deadspace. This is why this function does a lot of reformatting of the dataframe itself. The reason why this column ends up with the "Type of accommodation" label, is I think because "Type of accommodation" is the name of the columns index. Because of bad formatting it just ends up being the header of the LSOA column and the actual column header ends up in the first row of data.

helloaidank commented on asf_heat_pump_suitability/getters/get_target.py:
```diff
@@ -78,50 +90,128 @@ def get_df_target_property_type_uncensored() -> pl.DataFrame:
+        .rename({"Type of accomodation": "lsoa"})
+    )
+
+    # A small number of rows seem to erroneously have zero values for all property types, we need to remove them
```
What kind of number are we talking about here? Any idea why this might be the case?

crispy-wonton replied:
It was 4 rows in the target dataset (equivalent to 4 Data Zones). I'm not sure why this is the case but I assume this is an error in the NRS dataset as I checked, and it is present in the raw data.

helloaidank commented on asf_heat_pump_suitability/pipeline/reweight_epc/reweight_epc.py:
```diff
@@ -88,6 +88,8 @@ def generate_balance_sample(
         sample = sample.filter(~pl.col(feature).is_in(missing))
     lost_rows = len_before - len(sample)
 
+    # TODO generating dummies will fail and cause pipeline error if all rows are removed from sample in code above
```
What is to do on this? Raise an error when this occurs?

crispy-wonton replied:
See line below :)

`# TODO we need to check if len(sample) > 0 and only proceed with remaining code if it is`

I don't think it should error out. 

Basically there is an edge case where all the rows from the EPC sample for a specific LSOA are removed in the preprocessing for weighting. The case I found was a Scottish Data Zone which had only 1 EPC record left in the sample by the time it got to this stage of the reweighting pipeline.

At this point in the code where I left these comments, the code checks that the EPC subsample (in this case 1 row) only has categories that appear in the target (census) data for that LSOA. E.g. let's say the row has property type == flat. But in the census data maybe this LSOA has a count of 0 for flats. The current pipeline will remove all rows with flats from the EPC subsample, because we can't reweight something to 0 (because we are reweighting in multiple dimensions).

Therefore in this edge case, all rows are removed.

The next step is to check which categories are present in the target but missing from the sample. E.g. the target data may show the LSOA has 500 detached houses. Then the pipeline will append dummy rows with these missing variables to the subsample. E.g. it will try to add a dummy row with property type == detached house.

The error is actually caused in this edge case by this dummy generation. The code attempts to append a dummy row onto a dataframe, but as the dataframe has no data it fails.

Ultimately, we need to update it so that the pipeline recognises when this happens and then skips reweighting an LSOA in this case. If there are no rows, there is no data left to reweight anyway.

crispy-wonton replied:
I've rerun the `run_compute_epc_weights.py` script twice now in what I think are the same conditions as before and I can't recreate this bug. I don't know why.

I have looked for stochasticity in the script that would make this bug appear only sometimes, but I don't see any. I also don't think the source data has changed. I am going to leave the TODO note because it could still happen, but it can be fixed in a separate PR. I will create a new issue.

helloaidank commented on asf_heat_pump_suitability/pipeline/run_scripts/run_compute_epc_weights.py:
```diff
@@ -1,23 +1,26 @@
 """
-Add LSOA; MSOA data to EPC dataset and weight properties with Iterative Proportional Fitting per LSOA according to the
+Weight properties with Iterative Proportional Fitting per LSOA / Data Zone according to the
```
What is a data zone? Like an LA instead of an LSOA?

crispy-wonton replied:
Data Zone is our best equivalent of LSOA but for Scotland.

Data Zones:
> Composed of aggregates of Census Output Areas, data zones are large enough that statistics can be presented accurately without fear of disclosure and yet small enough that they can be used to represent communities. They are designed to have roughly standard populations of 500 to 1,000 household residents
[Source](https://spatialdata.gov.scot/geonetwork/srv/api/records/7d3e8709-98fa-4d71-867c-d5c8293823f2)

LSOAs:

> Lower layer Super Output Areas (LSOAs) are made up of groups of Output Areas (OAs), usually four or five. They comprise between 400 and 1,200 households and have a usually resident population between 1,000 and 3,000 persons.
[Source](https://www.ons.gov.uk/methodology/geography/ukgeographies/statisticalgeographies)

helloaidank commented on asf_heat_pump_suitability/pipeline/run_scripts/run_compute_epc_weights.py:
```diff
@@ -1,23 +1,26 @@
 - tenure (owner-occupied, social rental, private rental)
+- build year (pre- and post-1930 split, and unknown); [applies to England and Wales only*]
+
+*Data Zones in Scotland are reweighted on two features only (property type and tenure) due to the absence of target
```
Not sure this would be clear to someone coming across it for the first time, would likely need to spell it out more!

crispy-wonton replied:
Wasn't exactly sure what you mean but I've updated it to explain what Data Zones are. Let me know if it still needs further clarification.

helloaidank replied:
"Absence of target" would likely be unclear to quite a few people coming across the code. Might just need to be a bit clearer on why we only reweight on two features!

helloaidank commented on asf_heat_pump_suitability/pipeline/run_scripts/run_compute_epc_weights.py:
```diff
@@ -1,23 +1,26 @@
 
 To run:
-python -i asf_heat_pump_suitability/pipeline/run_scripts/run_compute_epc_weights.py --epc_path [path/to/unweighted/EPC] -y [YYYY] -q [N]
+python -i asf_heat_pump_suitability/pipeline/run_scripts/run_compute_epc_weights.py --epc [path/to/EPC] -y [YYYY] -q [N]
```
What is [N] here?

crispy-wonton replied:
N is meant to represent the integer for the quarter that you can input as an argument, but maybe it makes more sense to change it to Q!

crispy-wonton replied:
I have changed it to `Q` in PR 86!

helloaidank commented on asf_heat_pump_suitability/pipeline/run_scripts/run_compute_epc_weights.py:
```diff
@@ -80,18 +73,33 @@ def parse_arguments() -> argparse.Namespace:
-        epc_df = pl.read_csv(epc_path, columns=config["usecols"]["epc"])
-    else:
-        epc_df = pl.read_parquet(epc_path, columns=config["usecols"]["epc"])
+    epc_df = pl.read_parquet(
```
might be nice here to raise error if file is not parquet!

helloaidank commented on asf_heat_pump_suitability/pipeline/run_scripts/run_compute_epc_weights.py:
```diff
@@ -80,18 +73,33 @@ def parse_arguments() -> argparse.Namespace:
+        ],
+    )
 
     # Join ONSPD LSOA col
```
Does ONSPD stand for ONS Postcode Directory?

crispy-wonton replied:
Yes :)

helloaidank commented on asf_heat_pump_suitability/pipeline/run_scripts/run_compute_epc_weights.py:
```diff
@@ -80,18 +73,33 @@ def parse_arguments() -> argparse.Namespace:
+    )
 
     # Join ONSPD LSOA col
     epc_df = output_areas.standardise_col_postcode(epc_df, pcd_col="POSTCODE")
```
Maybe after standardisation you do a regex check to see if the postcode is in the correct format:
```
valid_postcode_mask = epc_df["POSTCODE"].str.contains(r"^[A-Z0-9 ]+$")
assert valid_postcode_mask.all(), "Some standardised postcodes contain invalid characters."
```

helloaidank commented on asf_heat_pump_suitability/pipeline/run_scripts/run_compute_epc_weights.py:
```diff
@@ -102,66 +110,79 @@ def parse_arguments() -> argparse.Namespace:
     # 1. Add standardised weighting feature columns to EPC and drop rows missing data required for reweighting
```
Potentially something like: 
`assert (epc_df["lsoa"].is_null().sum() / epc_df.shape[0]) < 0.05, "More than 5% of rows missing lsoa after ONSPD join, which is unexpected."`

helloaidank commented on asf_heat_pump_suitability/pipeline/run_scripts/run_compute_epc_weights.py:
```diff
@@ -102,66 +110,79 @@ def parse_arguments() -> argparse.Namespace:
+        logging.info(
+            f"Running reweighting for {key}. Reweighting using the following features: {features}"
+        )
+        epc_cleaned_df = epc_df.filter(pl.col("COUNTRY") == key)
```
`assert epc_cleaned_df.shape[0] > 0, f"No rows found for {key}, expected some data."`

helloaidank commented on asf_heat_pump_suitability/pipeline/run_scripts/run_compute_epc_weights.py:
```diff
@@ -102,66 +110,79 @@ def parse_arguments() -> argparse.Namespace:
+                weights["UPRN"].extend(_weights["UPRN"])
+                # Adding LSOA required for dummy rows
+                weights["lsoa"].extend([lsoa for i in range(len(_weights["UPRN"]))])
+                weights["weight"].extend(_weights["weight"])
```
Just a question more so for you, but maybe we can make an assert statement here for the weights to make sure that they all add up to 1 overall (if that makes sense?)

crispy-wonton replied:
Yes that makes sense. They should add to approximately 1. E.g. it might add to 1.00000001.

I have to think about this more because I don't think I want the pipeline to error out if one LSOA doesn't add up. But I've made a note of all your suggestions so that I can add tests in a separate PR. Thank you so much!

helloaidank reviewed (commented):
Hi Roisín,

Great piece of coding! 👍
I can imagine it must have been mildly annoying to create two separate sets of getter functions for EW and Scottish data, but you've done a good job. Thank you for clearly documenting your expectations in the PR—those guidelines were really helpful to follow! 👍

**Minor Suggestions:**

- Config File: A few tweaks to improve clarity. 
- Documentation: Some additional details could make the code easier to understand for others.
- Assertions: I’ve added a couple of suggestions.

**Output Review:**
I checked the output files and performed a sanity check on some of the results. Everything looks fine for both datasets! However, I noticed we lost a total of 23,801 rows, which averages out to about 0.5 rows per LSOA. This doesn’t seem significant but might be worth exploring further to understand why we're losing those rows. That said, it's more of a second/third-order concern at this stage.

**Performance:**
The processing times range from 0.05 seconds to 6.9 seconds, which seems reasonable. It's expected that some LSOAs take longer due to differences in their distribution.

- _**Reweighting has been applied correctly to each nation in the code. I.e. Scotland is reweighted appropriately on property_type and tenure features only, whereas England and Wales should also be reweighted on build_year**_ Checked this, seems to be good.
- **_Joins to make sure they are correct and we are not accidentally losing data we should keep. My intention is to retain all the rows that get assigned a weight, but not to keep the unweighted ones. The idea is to join this dataset back onto the full EPC dataset in a separate processing script._** I double checked the joins and the logic, it checks out
- **_The preprocessing steps in run_compute_epc_weights.py before reweighting occurs to ensure they are applied in the appropriate order before and after filtering for country_** Preprocessing steps seem to be in the correct order.
- **_Feel free to suggest places where we can add test/assertions to ensure the pipeline is executing the expected behaviours. I'd like to look at this properly myself later and add some in a separate PR._** I've added some suggestions!

Hope this is helpful!

crispy-wonton committed:
add TODO note in get_target.py to add Scottish nrooms target data if used later

crispy-wonton committed:
add notation to get_target.py for clarity

crispy-wonton committed:
update docstring at the top of run_compute_epc_weights.py to explain Data Zones

crispy-wonton committed:
update documentation in run_compute_epc_weights.py for clarity

crispy-wonton committed:
add assert statement to check EPC records exist for country key in run_compute_epc_weights.py

crispy-wonton commented:
@helloaidank thank you so much for your review and great suggestions.

I have taken notes of all your testing suggestions to add in a separate PR. I have also created issues #92 and #97 based on your review :) 

Otherwise, I think I have addressed all your comments. Please take a look and let me know if you're happy for me to merge :)

helloaidank commented:
Hi @crispy-wonton, thanks for taking the time to go through the comments, I've had a look and I'm happy! Happy with you creating issues for those improvements! Thanks again for the neat and tidy code. It's time to merge 🚀
